### PR TITLE
Implement deleteNode in smart contract

### PR DESCRIPTION
**Pull request by folex**

### CODE REVIEW

File: bootstrap/contracts/Deployer.sol [hunk 1]
```diff
@@ -301,13 +317,129 @@ contract Deployer {
...
+                }
+            }
+
+            // This should never happen if there's no bugs. But it's better to revert if there is some inconsistency.
+            require(j < nodeIDs.length, "error deleting node: nodeID wasn't found in nodeIDs");
+
+            // Check if that node is the last one hosting that app
+            if (nodeIDs.length == 1) {
+                // Remove app from clustersIds array and clusters mapping, and emit AppDeleted event of successful removal
+                removeApp(appID);
```
Comment (DieMyst) [inline bootstrap/contracts/Deployer.sol]: if the node is removed, we should find another one to work with the app. Am I right?
It is incorrect to delete an app when deleting a node.

File: bootstrap/contracts/Deployer.sol [hunk 2]
```diff
@@ -301,13 +317,129 @@ contract Deployer {
...
+                // Remove nodeID from nodeIDs array
+                if (j != nodeIDs.length - 1) {
+                    // Remove j-th node from nodeIDs replacing it by the last node in the array
+                    nodeIDs[j] = nodeIDs[nodeIDs.length - 1];
+                }
+                // Release the storage
+                delete nodeIDs[nodeIDs.length - 1];
+
+                // Decrease array length manually
+                assembly { mstore(nodeIDs, sub(mload(nodeIDs), 1)) }
```
Comment (DieMyst) [inline bootstrap/contracts/Deployer.sol]: I see in code something like this: `readyNodes.length--;` Is it the same, is it?

File: bootstrap/contracts/Deployer.sol [hunk 3]
```diff
@@ -301,13 +317,129 @@ contract Deployer {
...
+    function removeNodeFromAppsStorage(bytes32 nodeID, bytes32[] storage appIDsArray)
+        internal
+    {
+        for (uint i = 0; i < appIDsArray.length; i++) {
+            bytes32 appID = appIDsArray[i];
+            bytes32[] storage nodeIDs = apps[appID].cluster.nodeIDs;
+            uint nodeIDsLength = nodeIDs.length;
+
+            uint8 j = 0;
+            for (; j < nodeIDsLength; j++) {
```
Comment (DieMyst) [inline bootstrap/contracts/Deployer.sol]: this part is repeated, maybe move it to a method?

Comment (DieMyst) [reply bootstrap/contracts/Deployer.sol]: function `foundNodeIndex`

File: bootstrap/test/NodeDeleteTest.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            let receipt = await global.contract.deleteNode(nodeID, { from: anyone });
+            truffleAssert.eventEmitted(receipt, utils.nodeDeletedEvent, ev => {
+                return ev.id == nodeID;
+            });
+            return receipt;
+        }));
+
+        truffleAssert.eventEmitted(receipts.pop(), utils.appDeletedEvent);
+    });
+});
```
Comment (DieMyst) [inline bootstrap/test/NodeDeleteTest.js]: new line

File: bootstrap/contracts/Deployer.sol (hunk 3 above)
Comment (folex) [reply bootstrap/contracts/Deployer.sol]: Great idea! Done that

**Pull request merged**